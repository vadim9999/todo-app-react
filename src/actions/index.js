import { ADD_TASK, FOUND_WORD } from "../constants/action-types";


// export function addArticle(payload){
//     return { type: ADD_ARTICLE, payload}
// }


export function addTask(payload){
    console.log("Call Function in actions");
    console.log(payload);
    
    return { type: ADD_TASK, payload}
}

export function addFoundWord(payload){
    console.log("Found word");
    console.log(payload);
    
    return { type: FOUND_WORD, payload}
}

export function getData(){
    return fetch("https://jsonplaceholder.typicode.com/posts").then(
        response => response.json())
        .then(json => {
            return {
                type:"Data_LOADED",
                payload:json
            }
        })
    
}