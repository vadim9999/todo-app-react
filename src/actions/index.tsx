import { ADD_TASK, FOUND_WORD } from '../constants/action-types';

// export function addArticle(payload){
//     return { type: ADD_ARTICLE, payload}
// }

export function addTask(payload: any) {
  return { type: ADD_TASK, payload };
}

export function addFoundWord(payload: any) {
  return { type: FOUND_WORD, payload };
}

export const getTasks = (payload: any) => ({ type: 'GET_TASKS', payload });

export const deleteTaskById = (payload: any) => ({
  type: 'DELETE_TASK',
  payload
});

export const updateTaskById = (payload: any) => ({
  type: 'UPDATE_TASK',
  payload
});
// export function getData() {
//     return function(dispatch) {
//       return fetch("https://jsonplaceholder.typicode.com/posts")
//         .then(response => response.json())
//         .then(json => {
//             console.log("data");
//             console.log(json);

//           dispatch({ type: "DATA_LOADED", payload: json });
//         });
//     };
//   }

export const sortTasksByGrowthDate = (payload: any) => ({
  type: 'SORT_TASKS_BY_GROWTH_DATE',
  payload
});

export const sortTasksByDecreaseDate = (payload: any) => ({
  type: 'SORT_TASKS_BY_DECREASE_DATE',
  payload
});

export const addCurrentPage = (payload: any) => ({
  type: 'ADD_CURRENT_PAGE',
  payload
});

export const addSelectedRowKeys = (payload:any) =>({
  type:'ADD_SELECTED_ROW_KEYS',
  payload
})
export const addCurrentSelectedRowKeys = (payload: any) => ({
  type: 'ADD_CURRENT_SELECTED_ROW_KEYS',
  payload
});
// export function saveTask(){
//   return { type: "SAVE_TASK"}
// }
