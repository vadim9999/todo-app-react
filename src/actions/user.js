import { ADD_TASK, FOUND_WORD } from '../constants/action-types';

export const addUser = (payload) => ({
  type: 'ADD_USER',
  payload,
});

export const authenticate = (payload) => ({
  type: 'AUTHENTICATE', payload,
});

export const authorizate = (payload) => ({
  type: 'AUTHORIZATE', payload,
});
