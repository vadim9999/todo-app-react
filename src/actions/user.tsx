// import { ADD_TASK, FOUND_WORD } from '../constants/action-types';

export const addUser = (payload:any) => ({
  type: 'ADD_USER',
  payload,
});

export const authenticate = (payload:any) => ({
  type: 'AUTHENTICATE', payload,
});

export const authorizate = (payload:any) => ({
  type: 'AUTHORIZATE', payload,
});
