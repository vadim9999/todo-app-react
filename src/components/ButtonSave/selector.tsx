

export const getKeys = (selectedRowKeys:number[], 
  currentSelectedRowKeys:number[]) =>{

  const keysNotFounded = [];
    const changedRowKeys = [...currentSelectedRowKeys];

    for (let i = 0; i < selectedRowKeys.length; i++) {
      let founded = false;
      for (let j = 0; j < changedRowKeys.length; j++) {
        if (selectedRowKeys[i] === changedRowKeys[j]) {
          changedRowKeys.splice(j, 1);
          founded = true;
          break;
        }
      }

      if (!founded) {
        keysNotFounded.push(selectedRowKeys[i]);
      }
    }

    // changedRowKeys[...keysNotFounded]

    const result = changedRowKeys.concat(keysNotFounded);
    return result
}