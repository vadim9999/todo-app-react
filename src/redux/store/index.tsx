import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "redux-saga";
import mySaga from "../sagas/api-saga";
import tasksReducer from "../tasks/tasksSlice";

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: { tasks: tasksReducer },
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware().concat(sagaMiddleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

sagaMiddleware.run(mySaga);

