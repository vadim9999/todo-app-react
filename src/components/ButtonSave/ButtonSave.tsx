import React from 'react'
import { Button } from 'antd'

import moment from 'moment';
import { connect } from 'react-redux';
import { TasksTypes } from '../Interfaces';
import { updateTaskById } from '../../actions';

interface ButtonSaveProps {
  tasks: TasksTypes[];
  selectedRowKeys: number[];
  currentSelectedRowKeys: number[];
  user?: any;
  user_id: string;

  updateTaskById: any;
}

interface ButtonSaveState {
  loading: boolean;
}

const mapDispatchToProps = (dispatch: any) => {
  return {
    updateTaskById: (id: string) => dispatch(updateTaskById(id)),
  };
};

const mapStateToProps = (state: ButtonSaveProps) => {
  return {
    selectedRowKeys: state.selectedRowKeys,
    currentSelectedRowKeys: state.currentSelectedRowKeys,
    tasks: state.tasks,
    user_id: state.user._id
  };
};

export class ConnectedButtonSave extends React.Component<ButtonSaveProps, ButtonSaveState>{
  constructor(props: any) {
    super(props);

    this.state = {
      loading: false
    };
  }

  start = () => {
    const { tasks, selectedRowKeys, currentSelectedRowKeys } = this.props;

    this.setState({
      loading: true
    });

    const keysNotFounded = [];
    const changedRowKeys = [...currentSelectedRowKeys];

    for (let i = 0; i < selectedRowKeys.length; i++) {
      let founded = false;
      for (let j = 0; j < changedRowKeys.length; j++) {
        if (selectedRowKeys[i] === changedRowKeys[j]) {
          changedRowKeys.splice(j, 1);
          founded = true;
          break;
        }
      }

      if (!founded) {
        keysNotFounded.push(selectedRowKeys[i]);
      }
    }

    // changedRowKeys[...keysNotFounded]

    const rowKeys = changedRowKeys.concat(keysNotFounded);

    rowKeys.map((item: any) => {
      this.props.updateTaskById({
        ...tasks[item],
        // name: row.name,
        completed: !tasks[item].completed,
        login_id: this.props.user_id,
        date: moment().toISOString()
      });
    });

    setTimeout(() => {
      this.setState({
        loading: false
      });
    }, 1000);
  };

  render() {
    const { loading } = this.state;
    return (
      <Button
        type="primary"
        onClick={this.start}
        // disabled={equalRowKeys}
        loading={loading}
      >
        Save
          </Button>
    )
  }
}

const ButtonSave = connect(mapStateToProps,mapDispatchToProps)(ConnectedButtonSave)
export default ButtonSave