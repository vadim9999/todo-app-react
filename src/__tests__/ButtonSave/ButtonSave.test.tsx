import React from 'react'
import Enzyme, {mount, shallow} from 'enzyme'
import configureStore from 'redux-mock-store'
import Adapter from 'enzyme-adapter-react-16'
import {Provider} from 'react-redux'
import ButtonSave, {ConnectedButtonSave} from '../../components/ButtonSave/ButtonSave'
import {updateTaskById, addCurrentSelectedRowKeys, addSelectedRowKeys} from '../../actions'
import renderer from 'react-test-renderer'
import {createStore} from 'redux'
import {Button} from 'antd'
import { unmountComponentAtNode } from 'react-dom'
import rootReducer from '../../reducers'

Enzyme.configure({adapter: new Adapter()})

describe('Testing ButtonSave', ()=>{
  const initialState = {
    user: {
      _id:'1234'
    },
    tasks: [],
    currentPage: 1,
    selectedRowKeys: [],
    currentSelectedRowKeys: []
  }
  // const mockStore = configureStore()

  let store:any, wrapper:any;

  beforeEach(()=>{
    // store = mockStore(initialState)
    store = createStore(rootReducer,initialState)
    wrapper = mount(
      <Provider store ={store}>
        <ButtonSave />
      </Provider>
    )
  })

  // afterEach(()=>{
  //   wrapper.unmount()
  //   wrapper = null
  // })
  // it('checks on render button',()=>{
  //   expect(wrapper.find(ButtonSave).length).toEqual(1)
  // })

  // it('checks dispatching updateTaskById', () =>{
  //   let action
  //   store.dispatch(updateTaskById({
  //     name:'newtask',
  //     // name: row.name,
  //     completed: true,
  //     login_id: '1234',
  //     date: '12:22'
  //   }))

  //   // console.log(store.getState());
    
  //   // action = store.getActions()

  //   // expect(action[0].type).toBe('UPDATE_TASK')
  // })

  it('checks on click button without tasks',()=>{  
    const button = wrapper.find(Button)
    
    button.simulate('click')

  })

  it('checks on click when app has three tasks', ()=>{
    const button = wrapper.find(Button)

    store.dispatch({type:"GET_TASKS_SUCCESS", payload:{
      data:[
      {name:'task1',
      completed: true},
    {name:'task2',
    completed: false},
    {name:'task2',
    completed: false}
    ],
    selectedRowKeys: [0]
  }}
  )

    store.dispatch(addCurrentSelectedRowKeys([1,2]))
    
    // console.log(store);
    
    // wrapper.setProps({cord:"22"})
    
    wrapper.update()
    // jest.useFakeTimers()
    button.simulate('click')
  })

  describe('checks match initialState in props', ()=>{
    it('selectedRowKeys', ()=>{
      expect(wrapper.find(ConnectedButtonSave).prop('selectedRowKeys')).toEqual(initialState.selectedRowKeys)
    })
  
    it('currentSelectedRowKeys',() =>{
      expect(wrapper.find(ConnectedButtonSave).prop('currentSelectedRowKeys')).toEqual(initialState.currentSelectedRowKeys)
    })
  
    it('tasks',()=>{
      expect(wrapper.find(ConnectedButtonSave).prop('tasks')).toEqual(initialState.tasks)
    })

    it('user_id', ()=>{
      expect(wrapper.find(ConnectedButtonSave).prop('user_id')).toEqual(initialState.user._id)
    })
    // it('updateTaskById', ()=>{
    //   expect(wrapper.find(ConnectedButtonSave).prop('updateTaskById')({
    //     name:'newtask',
    //     // name: row.name,
    //     completed: true,
    //     login_id: '1234',
    //     date: '12:22'
    //   }))
    //   .toEqual(store.dispatch(updateTaskById({
    //     name:'newtask',
    //     // name: row.name,
    //     completed: true,
    //     login_id: '1234',
    //     date: '12:22'
    //   })))
    // })
  })
  
  
  describe('snapshot ButtonSave',()=>{
    it('checks snapshot',()=>{
      const renderValue = renderer.create(<ConnectedButtonSave {...initialState} />).toJSON()
      expect(renderValue).toMatchSnapshot()
    })
  })
})