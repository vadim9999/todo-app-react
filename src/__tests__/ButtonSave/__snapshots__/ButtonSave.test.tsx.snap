// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ButtonSave after authorization snapshot ButtonSave checks snapshot 1`] = `
<button
  className="ant-btn ant-btn-primary"
  onClick={[Function]}
  type="button"
>
  <span>
    Save
  </span>
</button>
`;

exports[`Testing ButtonSave snapshot ButtonSave checks snapshot 1`] = `
<button
  className="ant-btn ant-btn-primary"
  onClick={[Function]}
  type="button"
>
  <span>
    Save
  </span>
</button>
`;
